pvlib/modelchain.py

```diff
@@ -1158,39 +1353,69 @@ def prepare_inputs(self, weather):
                 self.system.get_irradiance,
                 self.tracking['surface_tilt'],
                 self.tracking['surface_azimuth'],
-                self.solar_position['apparent_zenith'],
-                self.solar_position['azimuth'])
+                self.results.solar_position['apparent_zenith'],
+                self.results.solar_position['azimuth'])
         else:
             self._prep_inputs_fixed()
             get_irradiance = partial(
                 self.system.get_irradiance,
-                self.solar_position['apparent_zenith'],
-                self.solar_position['azimuth'])
-
-        self.total_irrad = get_irradiance(
-            self.weather['dni'],
-            self.weather['ghi'],
-            self.weather['dhi'],
-            airmass=self.airmass['airmass_relative'],
+                self.results.solar_position['apparent_zenith'],
+                self.results.solar_position['azimuth'])
+
+        self.results.total_irrad = get_irradiance(
+            _tuple_from_dfs(self.weather, 'dni'),
+            _tuple_from_dfs(self.weather, 'ghi'),
+            _tuple_from_dfs(self.weather, 'dhi'),
+            airmass=self.results.airmass['airmass_relative'],
             model=self.transposition_model)
 
         return self
 
+    def _check_multiple_input(self, data, strict=True):
+        """Check that the number of elements in `data` is the same as
+        the number of arrays in `self.system`. If `strict` is False then
+        `data` does not have to be a tuple and length validation is not
+        performed. If `data` is a tuple of the correct length the indices
+        of each DataFrame it contains are compared for equality and an
+        error is raised if they differ.
+        """
+        if (not strict or self.system.num_arrays == 1) \
+                and not isinstance(data, tuple):
+            return
+        if strict and not isinstance(data, tuple):
+            raise ValueError("Input must be a tuple of length "
+                             f"{self.system.num_arrays}, "
+                             f"got {type(data).__name__}.")
```

Comment:
Should this be a `TypeError`?

Reply:
https://docs.python.org/3/library/exceptions.html#TypeError

"Passing arguments of the wrong type (e.g. passing a list when an int is expected) should result in a TypeError, but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a ValueError."

TypeError seems appropriate here.